File: quiz-backend/src/Topic.cpp
#include "Topic.h"

// Constructor that instantiates the proficiency for a user as well as the name of the topic
// The proficiency will be gathered from the DB
Topic::Topic(std::string name, int tID, int proficiency) : proficiency_(proficiency), topicID_(tID) topicName_(name) {}


std::string Topic::getName() const {
    return name_;
}

// Increment proficiency
void Topic::increment() {
    if (proficiency_ < kMaxProficiency) { //can decide on number later
        ++proficiency_;
    }
}

// Decrement proficiency
void Topic::decrement() {
    if (proficiency_ > 0) {
        --proficiency_;
    }
}

void Topic::activate() {
    active_ = true;
}

void Topic::deactivate() {
    active_ = false;
}

void Topic::addQuestion(const Question& question) {
        questions_.push_back(question);
    }

int Topic::getProficiency() const {
    return proficiency_;
}

bool Topic::isActive() const {
    return active_;
}

Question Topic::getRandomQuestion() const {
    std::srand(static_cast<unsigned int>(time(0)));
    int randIndex = std::rand() % questions_.size();
    return questions_[randIndex];
}

std::vector<Question> Topic::shuffleQuestionVector() {
    std::vector<Question> shuffledQuestions(questions_);
    std::random_shuffle(shuffledQuestions.begin(), shuffledQuestions.end());
    return shuffledQuestions;
}

int Topic::getTopicID() const {
    return topicID_;
}

bool Topic::containsQuestion(std::string q) {
    for (Question que : questions_) {
        if (que.getQuestion() == q) {
            return true;
        }
    }
    return false;
}

File: quiz-backend/src/models/Topic.py
import random
from models.Question import Question

class Topic:
    kMaxProficiency = 20

    def __init__(self, name, t_id, proficiency, active=True):
        self.topic_name = name
        self.topic_id = t_id
        self.proficiency = proficiency
        self.active = active
        self.questions = []

    def get_name(self):
        return self.topic_name

    def increment(self):
        if self.proficiency < self.kMaxProficiency:
            self.proficiency += 1

    def decrement(self):
        if self.proficiency > 0:
            self.proficiency -= 1

    def activate(self):
        self.active = True

    def deactivate(self):
        self.active = False

    def add_question(self, question):
        self.questions.append(question)

    def get_proficiency(self):
        return self.proficiency

    def is_active(self):
        return self.active

    def shuffle_question_vector(self):
        return random.sample(self.questions, len(self.questions))

    def get_topic_id(self):
        return self.topic_id

    def contains_question(self, q):
        return any(que.get_question() == q for que in self.questions)